File: backend/app/models/event.py
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, CreatedAt, UUIDPrimaryKey
from app.models.enums import EventStatus, pg_enum

if TYPE_CHECKING:
    from app.models.photo import Photo


class Event(UUIDPrimaryKey, CreatedAt, Base):
    __tablename__ = "events"

    title: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(220), unique=True, nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    venue: Mapped[str] = mapped_column(String(200), nullable=False)
    event_date: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    cover_image_url: Mapped[str | None] = mapped_column(String(500))
    status: Mapped[EventStatus] = mapped_column(
        pg_enum(EventStatus, "event_status"),
        nullable=False,
        default=EventStatus.DRAFT,
        server_default=EventStatus.DRAFT.value,
    )
    academic_year: Mapped[str] = mapped_column(String(20), nullable=False, index=True)

    photos: Mapped[list["Photo"]] = relationship(
        back_populates="event",
        cascade="all, delete-orphan",
        order_by="Photo.sort_order",
    )

File: client_app/lib/models/event.dart
import 'photo.dart';

class EventStatus {
  static const draft = 'draft';
  static const published = 'published';
}

class ClubEvent {
  ClubEvent({
    required this.id,
    required this.title,
    required this.slug,
    required this.description,
    required this.venue,
    required this.eventDate,
    required this.coverImageUrl,
    required this.status,
    required this.academicYear,
    required this.createdAt,
    this.photos,
  });

  final String id;
  final String title;
  final String slug;
  final String description;
  final String venue;
  final DateTime eventDate;
  final String? coverImageUrl;
  final String status;
  final String academicYear;
  final DateTime createdAt;

  /// Only present on `GET /events/{id}`, ordered by `sort_order`.
  final List<EventPhoto>? photos;

  bool get isPast => eventDate.isBefore(DateTime.now());

  factory ClubEvent.fromJson(Map<String, dynamic> json) {
    return ClubEvent(
      id: json['id'] as String,
      title: json['title'] as String,
      slug: json['slug'] as String,
      description: json['description'] as String,
      venue: json['venue'] as String,
      eventDate: DateTime.parse(json['event_date'] as String),
      coverImageUrl: json['cover_image_url'] as String?,
      status: json['status'] as String,
      academicYear: json['academic_year'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      photos: json['photos'] == null
          ? null
          : (json['photos'] as List)
              .map((p) => EventPhoto.fromJson(p as Map<String, dynamic>))
              .toList(),
    );
  }
}
